controllers/cache_handler.go

```diff
@@ -0,0 +1,497 @@
+			"newPhase", newNode.Status.Phase,
+			"oldPhase", oldNode.Status.Phase,
+		)
+	}
+
+	if cannotRecoverYet || (eventsToSend[chaosv1beta1.EventNodeRecoveredState] && len(eventsToSend) > 1) {
+		eventsToSend[chaosv1beta1.EventNodeRecoveredState] = false
+	}
+
+	return eventsToSend
+}
+
+// Cache life handler
+
+// createInstanceSelectorCache creates this instance's cache if it doesn't exist and attaches it to the controller
+func (r *DisruptionReconciler) manageInstanceSelectorCache(instance *chaosv1beta1.Disruption) error {
+	r.clearExpiredCacheContexts()
+
+	// remove check when StaticTargeting is defaulted to false
+	if instance.Spec.StaticTargeting == nil {
+		r.log.Debugw("StaticTargeting pointer is nil")
+	}
+
+	if instance.Spec.StaticTargeting == nil || *instance.Spec.StaticTargeting {
+		return nil
+	}
+
+	disNamespacedName := types.NamespacedName{Name: instance.Name, Namespace: instance.Namespace}
+	disSpecHash, err := instance.Spec.HashNoCount()
+
+	if err != nil {
+		return fmt.Errorf("error getting disruption hash")
+	}
+
+	disCacheHash := disNamespacedName.String() + disSpecHash
+	disCompleteSelector, err := targetselector.GetLabelSelectorFromInstance(instance)
+
+	if err != nil {
+		return fmt.Errorf("error getting instance selector: %w", err)
+	}
+	// if it doesn't exist, create the cache context to re-trigger the disruption
+	if _, ok := r.CacheContextStore[disCacheHash]; !ok {
+		// create the cache/watcher with its options
+		var cacheOptions k8scache.Options
+		if instance.Spec.Level == chaostypes.DisruptionLevelNode {
+			cacheOptions = k8scache.Options{
+				SelectorsByObject: k8scache.SelectorsByObject{
+					&corev1.Node{}: {Label: disCompleteSelector},
+				},
+			}
+		} else {
+			cacheOptions = k8scache.Options{
+				SelectorsByObject: k8scache.SelectorsByObject{
+					&corev1.Pod{}: {Label: disCompleteSelector},
+				},
+				Namespace: instance.Namespace,
+			}
+		}
+
+		cache, err := k8scache.New(
+			ctrl.GetConfigOrDie(),
+			cacheOptions,
+		)
+
+		if err != nil {
+			return fmt.Errorf("cache gen error: %w", err)
+		}
+
+		// attach handler to cache in order to monitor the cache activity
+		var info k8scache.Informer
+
+		if instance.Spec.Level == chaostypes.DisruptionLevelNode {
+			info, err = cache.GetInformer(context.Background(), &corev1.Node{})
+		} else {
+			info, err = cache.GetInformer(context.Background(), &corev1.Pod{})
+		}
+
+		if err != nil {
+			return fmt.Errorf("cache gen error: %w", err)
+		}
+
+		info.AddEventHandler(DisruptionTargetWatcherHandler{disruption: instance, reconciler: r})
+
+		// start the cache with a cancelable context and duration, and attach it to the controller as a watch source
+		ch := make(chan error)
+		cacheCtx, cacheCancelFunc := context.WithTimeout(context.Background(), instance.Spec.Duration.Duration()+*r.ExpiredDisruptionGCDelay*2)
+
+		go func() { ch <- cache.Start(cacheCtx) }()
+
+		r.CacheContextStore[disCacheHash] = CtxTuple{cacheCtx, cacheCancelFunc}
```

Comment:
Should we monitor the cache size? I know that this code was already existing before so eventually something to do in another PR.

Reply:
@nathantournant Did add some metrics regarding the cache, notably: how many caches are still in the cache array, I think?

Reply:
Currently there's a cache-array gauge metric to make sure the number of caches is never durably superior to the number of ongoing disruptions. It's currently reported [with the recurrent metrics](https://github.com/DataDog/chaos-controller/blob/4eca165f072f00c54ca008f8434dcad51b99ef38/controllers/disruption_controller.go#L1505)
Adding a count on cache addition could help for debugging but I believe a debugging-level log line would be more useful for this
Otherwise I wouldn't exactly know how to monitor an individual cache/memory consumption, but would be happy to work on it with @clairecng

Reply:
Oops missed that one, I think we are good for now then!